### DISCUSSION TITLE
Using OpenVDB to detect gaps/holes/constricted regions in an SDF

### DISCUSSION THREAD

**Discussion by NiranAGMT** [AcademySoftwareFoundation/openvdb#1468] [Q&A] [answered]
I want to implement an algorithm that detects constricted regions (such as holes/gaps) of a signed-distance-field (SDF), where the zero levelset corresponds to the surface of some CAD geometry. I was recommended to use the levelset advection and boolean operation tools of OpenVDB for this purpose. The general idea I had in mind was as follows:

1. Supposing the size threshold for gaps is `ds`, then identify the levelset iso-surface corresponding to the distance `ds` in the SDF.
2. Uniformly advect this iso-surface backward (in the direction of the zero-levelset) by a distance of `ds`.
3. Apply some boolean operation between the iso-surface from 2. and the zero-levelset iso-surface to extract regions between the two iso-surfaces that were not 'swept over' by the iso-surface from 2.

Note that I have not quite figured out the exact boolean operations in 3.

Does anyone have any experience with identifying gaps/holes using OpenVDB, with a method similar to the above? If so, any advice on which OpenVDB tools I should use would be very much appreciated.

**Comment by apradhana** [comment, marked as answer]
Hello,
One way to solve this problem is by combining the trick of (a) doing morphological closing, and (b) computing the difference between your original geometry with the result of doing morphological closing (from (a)). To do part (a): you dilate the level-set, you reinitialize the level-set, then you erode. The result of this will fill the holes in your level-set. By doing a difference with the original level-set, you will detect the holes.

**Reply by NiranAGMT** [reply]
Thanks so much! That's a really elegant solution!
